apigateway/src/main/kotlin/ru/zveron/apigateway/registry/ProtoDefinitionRegistry.kt

```diff
@@ -0,0 +1,96 @@
+package ru.zveron.apigateway.registry
+
+import com.google.protobuf.DescriptorProtos
+import com.google.protobuf.Descriptors.FileDescriptor
+import io.grpc.MethodDescriptor
+import io.grpc.kotlin.ClientCalls
+import io.grpc.protobuf.lite.ProtoLiteUtils
+import io.grpc.reflection.v1alpha.ServerReflectionRequest
+import io.grpc.reflection.v1alpha.ServerReflectionResponse
+import kotlinx.coroutines.reactive.awaitSingle
+import mu.KLogging
+import org.springframework.cloud.netflix.eureka.reactive.EurekaReactiveDiscoveryClient
+import org.springframework.stereotype.Component
+import java.util.concurrent.ConcurrentHashMap
+
+@Component
+class ProtoDefinitionRegistry(
+    private val channelRegistry: GrpcChannelRegistry,
+    private val eurekaClient: EurekaReactiveDiscoveryClient,
+) {
+
+    companion object : KLogging()
+
+    private val reflectionMethodDescr = MethodDescriptor.newBuilder<ServerReflectionRequest, ServerReflectionResponse>()
+        .setFullMethodName(
+            MethodDescriptor.generateFullMethodName(
+                "grpc.reflection.v1alpha.ServerReflection",
+                "ServerReflectionInfo"
+            )
+        )
+        .setRequestMarshaller(ProtoLiteUtils.marshaller(ServerReflectionRequest.getDefaultInstance()))
+        .setResponseMarshaller(ProtoLiteUtils.marshaller(ServerReflectionResponse.getDefaultInstance()))
+        .setType(MethodDescriptor.MethodType.UNARY)
+        .build()
+
+    private val serviceToFileDescriptorMap = ConcurrentHashMap<String, FileDescriptor>()
+
+    private val serviceToProtoFile = ConcurrentHashMap<String, String?>()
+
+    suspend fun getProtoFileDescriptor(serviceName: String, grpcServiceName: String) =
+        serviceToFileDescriptorMap["$grpcServiceName-$serviceName"] ?: createProtoFileDescriptor(
+            grpcServiceName,
+            serviceName
+        )
+
+    private suspend fun createProtoFileDescriptor(
+        protoServiceName: String,
+        serviceName: String,
+    ): FileDescriptor {
+        val protoFileName = serviceToProtoFile["$protoServiceName-$serviceName"]
+            ?: eurekaClient.getInstances(serviceName)
+                .awaitSingle()?.metadata?.get(protoServiceName)
```

Comment:
вот я  5 минут тупила на этой строчке, типа, зачем тебе по grpcServiceName вытаскивать из метаданных

не проще тогда вместо: 
 ```
eureka:
  instance:
    metadata-map:
      BlacklistService: 'crud.proto'
```
заменить название поля с `BlacklistService `на `contract`? 

Так будет меньше проебов и согласованности везде

Reply:
и тогда будет код выглядеть так(ну, вынесенная в константу, само собой): 
```suggestion
                .awaitSingle()?.metadata?.get("contract")
```

Reply:
не будет, тк у сервиса может быть несколько грпц сервисов